feat: Add support for Dart 3.1, remove support for Dart 2.18

mbfakourii opened the pull request:
## Pull Request

- Report security issues [confidentially](https://github.com/parse-community/Parse-SDK-Flutter/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/Parse-SDK-Flutter/blob/master/LICENSE).
- Link this pull request to an [issue](https://github.com/parse-community/Parse-SDK-Flutter/issues?q=is%3Aissue).

## Issue
<!-- Add the link to the issue that this PR closes. -->

Closes: #968

## Approach
Improve support policy and CI

## Tasks
<!-- Delete tasks that don't apply. -->

- [ ] Add tests
- [ ] Add changes to documentation (guides, repository pages, code comments)

mtrezza commented on packages/dart/README.md:
```diff
@@ -28,26 +30,34 @@ This library gives you access to the powerful Parse Server backend from your Dar
-To install, add the Parse Dart SDK as a [dependency](https://pub.dev/packages/parse_server_sdk/install) in your `pubspec.yaml` file.
+To install, add the Parse Dart SDK as
+a [dependency](https://pub.dev/packages/parse_server_sdk/install) in your `pubspec.yaml` file.
 
 ## Documentation
 
 Find the full documentation in the [Parse Dart SDK guide][guide].
 
 ## Contributing
 
-We want to make contributing to this project as easy and transparent as possible. Please refer to the [Contribution Guidelines](https://github.com/parse-community/Parse-SDK-Flutter/blob/master/CONTRIBUTING.md).
+We want to make contributing to this project as easy and transparent as possible. Please refer to
+the [Contribution Guidelines](https://github.com/parse-community/Parse-SDK-Flutter/blob/master/CONTRIBUTING.md).
```
There seems to be some issue with your automated code formatting (if you haven't made these changes manually). There should not be any line breaks, these are added automatically as needed by the markdown renderer.

mbfakourii replied:
Yes, it changed, please check again

mtrezza replied:
looks good

mbfakourii commented on .github/workflows/ci.yml:
```diff
@@ -15,24 +15,25 @@ jobs:
         include:
-          # Dart framework may contain breaking changes in minor version releases, not following semver.
-          - name: Dart 3.0, Ubuntu
+          # Dart 3.1
```
I think the main reason is the resource limitation from github for actions.

As you know, running on non-Linux operating systems requires more [resources](https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration#usage-limits).

In general, there is no need to test all versions on all operating systems because the reason for this testing is not to evaluate the internal package structure. The reason for this action is to examine package execution, development constraints, platform limitations, and local variables.

mtrezza replied:
Could it be that these "development constraints, platform limitations, and local variables" change depending on the combination of Dart version and OS platform? In other words, is it possible that the Parse SDK works with Dart 3.0 on Windows but not with Dart 3.1 on Windows?

mbfakourii replied:
No, nothing like that happens and there is no problem.

mtrezza replied:
I'll try this out real quick in https://github.com/parse-community/Parse-SDK-Flutter/actions/runs/6398999438 and revert if all tests pass.

mtrezza commented on packages/dart/README.md:
```diff
@@ -50,4 +52,5 @@ Find the full documentation in the [Parse Dart SDK guide][guide].
 We want to make contributing to this project as easy and transparent as possible. Please refer to the [Contribution Guidelines](https://github.com/parse-community/Parse-SDK-Flutter/blob/master/CONTRIBUTING.md).
 
 [guide]: https://docs.parseplatform.org/dart/guide/
+
 [open-collective-link]: https://opencollective.com/parse-server
```
```suggestion
[guide]: https://docs.parseplatform.org/dart/guide/
[open-collective-link]: https://opencollective.com/parse-server

```

mtrezza commented on packages/flutter/README.md:
```diff
@@ -50,4 +53,5 @@ Find the full documentation in the [Parse Flutter SDK guide][guide].
 We want to make contributing to this project as easy and transparent as possible. Please refer to the [Contribution Guidelines](https://github.com/parse-community/Parse-SDK-Flutter/blob/master/CONTRIBUTING.md).
 
 [guide]: https://docs.parseplatform.org/flutter/guide/
+
 [open-collective-link]: https://opencollective.com/parse-server
```
```suggestion
[guide]: https://docs.parseplatform.org/flutter/guide/
[open-collective-link]: https://opencollective.com/parse-server
```

mtrezza commented on packages/dart/README.md:
```diff
@@ -30,12 +30,14 @@ This library gives you access to the powerful Parse Server backend from your Dar
 
 The Parse Dart SDK is continuously tested with the most recent release of the Dart framework to ensure compatibility. To give developers time to upgrade their app to the newest Dart framework, previous Dart framework releases are supported for at least 1 year after their [release date](https://dart.dev/get-dart/archive).
```
```suggestion
The Parse Dart SDK is continuously tested with the most recent release of the Dart framework to ensure compatibility. To give developers time to upgrade their app to a newer Dart framework, previous Dart framework releases are supported for at least 1 year after the [release date](https://dart.dev/get-dart/archive) of the next higher significant version.

```

mtrezza commented on packages/dart/README.md:
```diff
@@ -28,14 +28,17 @@ This library gives you access to the powerful Parse Server backend from your Dar
+| Dart 3.1  | 3.1.2          | Sep 2024       | ✅ Yes      |
+| Dart 3.0  | 3.0.7          | May 2024       | ✅ Yes      |
+| Dart 2.19 | 2.19.6         | Mar 2024       | ✅ Yes      |
+| Dart 2.18 | 2.18.5         | Nov 2023       | ❌ No       |
+
+> Parse Dart SDK requires **Dart >=2.19.6**
```
This extra line is covered by the table above.

```suggestion
| Dart 2.18 | 2.18.5         | Nov 2023       | ❌ No       |

```

mtrezza commented on packages/dart/README.md:
```diff
@@ -28,14 +28,16 @@ This library gives you access to the powerful Parse Server backend from your Dar
+| Dart 3.1  | 3.1.2          | Sep 2024       | ✅ Yes      |
+| Dart 3.0  | 3.0.7          | May 2024       | ✅ Yes      |
+| Dart 2.19 | 2.19.6         | Mar 2024       | ✅ Yes      |
+| Dart 2.18 | 2.18.5         | Nov 2023       | ❌ No       |
```
```suggestion

```

mtrezza commented on packages/dart/README.md:
```diff
@@ -28,14 +28,14 @@ This library gives you access to the powerful Parse Server backend from your Dar
+|-----------|----------------|----------------|------------|
+| Dart 3.1  | 3.1.2          | Sep 2024       | ✅ Yes      |
+| Dart 3.0  | 3.0.7          | May 2024       | ✅ Yes      |
+| Dart 2.19 | 2.19.6         | Mar 2024       | ✅ Yes      |
```
```suggestion
| Dart 2.19 | 2.19.6         | Mar 2024       | ✅ Yes      |
```

mtrezza commented on packages/dart/README.md:
```diff
@@ -28,14 +28,14 @@ This library gives you access to the powerful Parse Server backend from your Dar
+|-----------|----------------|----------------|------------|
+| Dart 3.1  | 3.1.2          | Sep 2024       | ✅ Yes      |
+| Dart 3.0  | 3.0.7          | May 2024       | ✅ Yes      |
+| Dart 2.19 | 2.19.6         | Mar 2024       | ✅ Yes      |
 
```
```suggestion
| Dart 2.19 | 2.19.6         | Mar 2024       | ✅ Yes      |

```

mbfakourii replied:
According the removal of the outdated version, I think it is better to remove the Compatible tab

mtrezza replied:
The tab has a different purpose, it's a standard tab in all our repositories.

mtrezza commented on packages/flutter/README.md:
```diff
@@ -30,12 +30,12 @@ This library gives you access to the powerful Parse Server backend from your Flu
 
 The Parse Flutter SDK is continuously tested with the most recent release of the Flutter framework to ensure compatibility. To give developers time to upgrade their app to the newest Flutter framework, previous Flutter framework releases are supported for at least 1 year after their [release date](https://docs.flutter.dev/release/archive?tab=linux). The Parse Flutter SDK depends on the Parse Dart ...
```
```suggestion
The Parse Flutter SDK is continuously tested with the most recent release of the Flutter framework to ensure compatibility. To give developers time to upgrade their app to a newer Flutter framework, previous Flutter framework releases are supported for at least 1 year after the [release date](https://docs.flutter.dev/release/archive?tab=linux) of the next higher significant version. The Parse Flutter SDK depends on the Parse Dart SDK which may require a higher Dart framework version than the Flutter framework version, in which case a Flutter framework version cannot be supported.
```

mtrezza reviewed (commented):
Looks good; changed this to `feat` PR, since it's adding Dart 3.1 support; added breaking label since it's removing Dart 2.18 support. This will publish as Parse Dart SDK 6.0.

Is this ready for merge?

mbfakourii commented:
> Looks good; changed this to `feat` PR, since it's adding Dart 3.1 support; added breaking label since it's removing Dart 2.18 support. This will publish as Parse Dart SDK 6.0.
> 
> Is this ready for merge?

It seems good

mtrezza commented:
Could you add the Dart version change?

I just realized this PR has changes for Dart and Flutter, but I think our process so far has been to keep them separate, as they are 2 separate packages.

So this is also a breaking change for the Parse Flutter SDK?

mtrezza commented:
This PR is more than a refactor, I've explained in https://github.com/parse-community/Parse-SDK-Flutter/pull/969#pullrequestreview-1656281916 why.

We haven't done multi-package changes in a single PR so far. That's something new and we'd need to explore how that affects our current process. I'd rather see these PRs moving forward that being stuck unnecessarily.

mbfakourii commented:
> This PR is more than a refactor, I've explained in [#969 (review)](https://github.com/parse-community/Parse-SDK-Flutter/pull/969#pullrequestreview-1656281916) why.
> 
> We haven't done multi-package changes in a single PR so far. That's something new and we'd need to explore how that affects our current process. I'd rather see these PRs moving forward that being stuck unnecessarily.

Done

mbfakourii commented:
The related errors are due to the marge between Flutter and Dart versions, which will be resolved in the next PR

mtrezza reviewed (approved):
Looks good!

mtrezza reviewed (changes requested):
CI fails with error:

```
[SEVERE] mockito:mockBuilder on test/test_utils.dart:

This builder requires Dart inputs without syntax errors.
However, asset:parse_server_sdk/test/test_utils.dart (or an existing part) contains the following errors.
test_utils.dart:116:5: Expected an identifier.
test_utils.dart:116:5: Expected to find ';'.

Try fixing the errors and re-running the build.
```

mbfakourii commented:
> CI fails with error:

It is due to the interference of two packages.

mtrezza commented:
How will it be fixed?

mtrezza commented on packages/dart/CHANGELOG.md:
```diff
@@ -1,3 +1,13 @@
+## [6.0.0](https://github.com/parse-community/Parse-SDK-Flutter/compare/dart-5.1.3...flutter-6.0.0) (2023-07-18)
```
```suggestion
## [6.0.0](https://github.com/parse-community/Parse-SDK-Flutter/compare/dart-5.1.3... dart-6.0.0) (2023-10-09)
```

mtrezza commented on packages/dart/CHANGELOG.md:
```diff
@@ -1,3 +1,13 @@
+## [6.0.0](https://github.com/parse-community/Parse-SDK-Flutter/compare/dart-5.1.3... dart-6.0.0) (2023-10-09)
```
```suggestion
## [6.0.0](https://github.com/parse-community/Parse-SDK-Flutter/compare/dart-5.1.3...dart-6.0.0) (2023-10-09)
```

mtrezza reviewed (commented):
The changes look good, but the CI needs to pass as well to merge; do you know what the issue here is and how it can be fixed?

mtrezza commented:
Thanks, so the Flutter tests are failing because the Parse Flutter SDK uses the unreleased Parse Dart SDK in this repository?

mbfakourii commented:
> Thanks, so the Flutter tests are failing because the Parse Flutter SDK uses the unreleased Parse Dart SDK in this repository?

Yes, that's correct.

mtrezza commented:
Should we change that? It doesn't seem to make sense if one package is released independently from the other?

mbfakourii commented:
> I think we have to change that, because merging a PR despite failing tests seems counter-intuitive. It's also taking more time to review a PR, because we always have to look into why a job is failing. We also have several other PR blocked because their tests are failing until this PR is merged.
> 
> I believe as long as we don't modify and release both packages with 1 PR (like you had it earlier in this PR, but we need to check the implications), the Flutter tests should use the published Parse Dart SDK, not the local package. What do you think?

It is a good suggestion to use the [paths](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore) in Github action for this issue

mtrezza commented:
Do you mean we could use paths to run exclude the Flutter CI jobs when the flutter files have not been modified?

mbfakourii commented:
> Do you mean we could use paths to exclude the Flutter CI jobs when the flutter files have not been modified?

No, I mean that the implementation of CI is created based on the changes of a certain folder, for example, we have two ci, one for dart with the package/dart path and one for flutter with the package/flutter path, which can be executed by changing the corresponding path files.

mtrezza commented:
We'd have to look into that. I think the issue here is that if someone downloads the repository and runs Flutter tests, they would still fail locally. So we would end up with "half working" releases; with no effect for the developer, but maybe making it difficult for the contributor.

How difficult would it be to make the Flutter SDK use the published Parse Dart SDK instead of the local one?

mbfakourii commented:
> We'd have to look into that. I think the issue here is that if someone downloads the repository and runs Flutter tests, they would still fail locally. So we would end up with "half working" releases; with no effect for the developer, but maybe making it confusing for the contributor.
> 
> How difficult would it be to make the Flutter SDK use the published Parse Dart SDK instead of the local one?

My suggestion is to add range support to Flutter. Something like this:

```parse_server_sdk: '>=3.0.0 <6.0.0'```

This would allow Flutter package to support a range of versions, such as 3 to 6.

mtrezza commented:
Not sure this would be possible. These are major versions and there may be breaking changes. How about using the latest published version but keep the major version fixed? Like a range of `parse_server_sdk: '>=3.0.0 <4.0.0'`

mbfakourii commented:
> Not sure this would be possible. These are major versions and there may be breaking changes. We already have a range with `parse_server_sdk: ^5.1.2`, the flutter SDK just needs to use the published version; not the local version.

We currently do not use the local version.
The reason for the CI error is related to the removal of support from version 2.18

mtrezza commented:
How is this possible if we did not actually remove Dart 2.18 support yet? It's only removed in this PR, not in the published Parse Dart SDK version.

mbfakourii commented:
I think I found the problem.
The issue here is that we are [checking](https://github.com/parse-community/Parse-SDK-Flutter/blob/dc7e3d6a28eab84b1ec00409ae62c18a7c06ddf0/.github/workflows/ci.yml#L120) the Parse Dart package in the Flutter section of CI.

mtrezza commented:
Makes sense, can we remove this line from the CI?

mbfakourii commented:
> Makes sense, can we remove this line from the CI?

I think to remove this item, it is better to do [PR flutter](https://github.com/parse-community/Parse-SDK-Flutter/pull/971).

What is your opinion about the separation of two CI (flutter and dart)?

mtrezza commented:
Sure, so should the line be removed in https://github.com/parse-community/Parse-SDK-Flutter/pull/971#pullrequestreview-1674817541?

mbfakourii commented:
@mtrezza 

What do you think about [melos](https://pub.dev/packages/melos)?

mtrezza commented:
Would it be enough to do https://github.com/parse-community/Parse-SDK-Flutter/pull/969#issuecomment-1760099830?

mbfakourii commented:
> I think you already identified the solution in [#969 (comment)](https://github.com/parse-community/Parse-SDK-Flutter/pull/969#issuecomment-1759782645). If you just remove the checking in the flutter section of the CI in the [#971 (review)](https://github.com/parse-community/Parse-SDK-Flutter/pull/971#pullrequestreview-1674817541) PR, then all these changes of splitting up the CI are unnecessary. We also want to keep the CI files as uniform as possible across repositories, for easier maintenance, so creating 2 files here if not necessary should be avoided if possible.

But didn't you say to separate it [here](https://github.com/parse-community/Parse-SDK-Flutter/pull/969#issuecomment-1760099830)?

mtrezza commented:
Yes, remove the line in the flutter PR https://github.com/parse-community/Parse-SDK-Flutter/pull/971, so we can merge the flutter PR first, then rebase this PR and there won't be any issues anymore, right?

mbfakourii commented:
> Yes, remove the [line](https://github.com/parse-community/Parse-SDK-Flutter/blob/dc7e3d6a28eab84b1ec00409ae62c18a7c06ddf0/.github/workflows/ci.yml#L120) in the flutter PR #971 (or adapt the CI so that Dart is not tested when running the flutter CI, without using `path` in the CI), so we can merge the flutter PR first, then rebase this PR and there won't be any issues anymore, right?

Done, please check again

mtrezza commented on .github/workflows/ci.yml:
```diff
@@ -16,47 +15,47 @@ jobs:
         run: dart pub get --directory packages/dart
       - name: Run build_runner
-        run: (cd packages/dart && dart run build_runner build --delete-conflicting-outputs)
+        run: cd packages/dart && dart run build_runner build --delete-conflicting-outputs
```
```suggestion
        run: (cd packages/dart && dart run build_runner build --delete-conflicting-outputs)
```

mtrezza commented on .github/workflows/ci.yml:
```diff
@@ -16,47 +15,47 @@ jobs:
         run: cd packages/dart && dart pub publish --dry-run
       - name: Run tests
-        run: (cd packages/dart && dart test --coverage=coverage)
+        run: cd packages/dart && dart test --coverage=coverage
```
```suggestion
        run: (cd packages/dart && dart test --coverage=coverage)
```

mtrezza commented on packages/dart/CHANGELOG.md:
```diff
@@ -1,3 +1,13 @@
+## [6.0.0](https://github.com/parse-community/Parse-SDK-Flutter/compare/dart-5.1.3...dart-6.0.0) (2023-10-09)
```
```suggestion
## [6.0.0](https://github.com/parse-community/Parse-SDK-Flutter/compare/dart-5.1.3...dart-6.0.0) (2023-10-16)
```

mtrezza reviewed (dismissed):
Looks good, waiting for CI to pass...

mtrezza commented on .github/workflows/ci.yml:
```diff
@@ -16,31 +15,31 @@ jobs:
+            sdk: 3.1.2
           # Only the latest Dart framework version (above) is tested with all architectures. Previous
           # Dart framework versions (below) are only tested with Ubuntu to reduce CI resource usage.
+          - name: Dart 3.0
```
```suggestion
          - name: Dart 3.0, Ubuntu
```
